### ISSUE TITLE
Child Jobs not executing Successful Completion callback

### ISSUE THREAD

**Issue by DaveWhitmer** [mperham/sidekiq#3859]
I am having an issue where jobs appear to not be reporting their completion or success status back to the parent batch.  There are no errors in the logs and Redis has plenty of memory.  This is an isolated application and does not share Redis.

![image](https://user-images.githubusercontent.com/4665368/40930695-6949776a-67f6-11e8-9a8b-3bd788dbc7fc.png)

In the attached screenshot, you can see that I kept manually rerunning the same parent job until I got all of the child jobs to complete on the final attempt at the top.  This problem is intermittent and I can resolve it by rerunning the parent job.  I can confirm that the last action in the child job is completing successfully and the job is removed from the queue and as a busy process.  It just appears that it is intermittently not executing the callback.

This was occurring with Rails 5.0.0, upon discussion over email, I have upgraded to Rails 5.0.7 and will update in a few days if this appears to resolve the issue.

Ruby version:  2.3.3
Sidekiq / Pro / Enterprise version(s): Pro-3.7.1

Please include your initializer and any error message with the full backtrace.
None

Are you using an old version?  No
Have you checked the changelogs to see if your issue has been fixed in a later version?  N/A

https://github.com/mperham/sidekiq/blob/master/Changes.md
https://github.com/mperham/sidekiq/blob/master/Pro-Changes.md
https://github.com/mperham/sidekiq/blob/master/Ent-Changes.md

**Comment by mperham** [comment]
Could you include your sidekiq initializer code?  Would it be possible to share the code that is creating the batch(es) in question?

**Comment by DaveWhitmer** [comment]
I am not using a Sidekiq intializer.  Just out of the box.

The code is below.  I have thrown all kinds of checks on this to try and figure out what is causing this problem and put it all in one class.  But this is the current version and has the same problem as the much prettier versions.  Good thing is you can see everything here in this one class.  This rebuilds giant ACH files in a multi-threaded manner with 30,000+ transactions.  

If you are not familiar with ACH files, the problem is that each batch has a check record and the entire file has a separate check record as well, so until each batch is assembled with released entries and all the batches are present, you can't calculate the check records.  I could cheat and do it from the database, but it really needs to check the actual records that are going to be written to the text file.  

My solution is to store each batch in a Redis value in parallel then assemble them through a serial process.  Because this is putting together entries from a bigger ACH file and only releasing part, I have to know which batches were used in this "Rebatching", so I know have an ugly join table name in here.

This is fairly easy to do in a serial fashion, but takes a VERY long time.  My problem is that everything runs perfect on a small number of batches, but when I have a larger number of batches and those batches have a large number of entries, this is where I tend to see what appears the callback failing to fire to either on_success or on_complete.

All of the batches will be correctly joined to the rebatch and all of the text records will be present in Redis.  Sorry for the ugly code....

```

class AchReBatchAssembler
  include Sidekiq::Worker
  sidekiq_options queue: "high", backtrace: true

  def perform(batch_number, batch_id, ach_re_batch_id)
    batch_create(batch_number, batch_id, ach_re_batch_id)
  end

  def on_success(status, options)
    rebatch_assemble(options['pid'], options['file_count'])
  end

  def on_complete(status, options)
    rebatch = AchReBatch.find(options['pid'])
    batch_ids = rebatch.entries.where(transmitted: false).pluck(:batch_id)
    message = 'Failed VLB Batches: ' + batch_ids
    user = User.find_by(email: 'XXXX@XXXXXXX.com')
    user.send_notification_email(message)
    rebatch.ach_re_batch_batches.update_all processed: false
    AchReBatchAssembler.assemble(options['pid'], options['file_count']) if status.failures != 0
  end

  def self.assemble(ach_re_batch_id, file_count)
    rebatch = AchReBatch.find(ach_re_batch_id)
    $redis.del("Batches ID: #{rebatch.id}")

    batches = Batch.joins(:entries).where(entries: { ach_re_batch: rebatch }).distinct.order(batch_count: :asc)
    rebatch.credits_cents = rebatch.entries.where(transaction_type: "CREDIT").sum(:amount_cents)
    rebatch.debits_cents = rebatch.entries.where(transaction_type: "DEBIT").sum(:amount_cents)
    rebatch.batch_count = batches.count
    rebatch.prenote_count = rebatch.entries.where('transaction_type = ? OR transaction_type = ?', "PRENOTE CREDIT", "PRENOTE DEBIT").count
    rebatch.entry_count = rebatch.entries.count
    rebatch.save
    batch_number = 1
    parent = Sidekiq::Batch.new
    parent.on(:success, self, { 'pid' => ach_re_batch_id, 'file_count' => file_count })
    parent.jobs do
      batches.each do |batch|
        rebatch.batches << batch rescue ActiveRecord::RecordNotUnique
        perform_async(batch_number, batch.id, rebatch.id)
        batch_number += 1
      end
    end
    ReBatchFileCheck.perform_in(10.minutes, ach_re_batch_id, file_count)
  end

  def batch_create(batch_number, batch_id, ach_re_batch_id)
    ach_re_batch = AchReBatch.find_by_id(ach_re_batch_id)
    batch = Batch.find_by_id(batch_id)
    ach_re_batch_batch = AchReBatchBatch.find_by(ach_re_batch: ach_re_batch, batch: batch)
    sec = batch.standard_entry_class_code
    scc = batch.service_class_code
    sd = batch.settlement_date.strftime('%j')
    originator = batch.originator
    originating_dfi = batch.originating_dfi_identification
    batch_entry_addenda_count = 0
    credit_check = 0
    debit_check = 0

    case sec
    when "IAT"
      lines = "5#{scc}"
      lines += ' ' * 16
      lines += (batch.foreign_exchange_indicator.nil? ? ' ' * 2 : batch.foreign_exchange_indicator.ljust(2, ' '))
      lines += (batch.foreign_exchange_reference_indicator.nil? ? ' ' * 1 : batch.foreign_exchange_reference_indicator.ljust(1, ' '))
      lines += (batch.foreign_exchange_reference.nil? ? ' ' * 15 : batch.foreign_exchange_reference.ljust(15, ' '))
      lines += 'US'
      lines += (batch.originator_identification.nil? ? ' ' * 10 : batch.originator_identification.ljust(10, ' '))
      lines += sec
      lines += (batch.company_entry_description.nil? ? ' ' * 10 : batch.company_entry_description.ljust(10, ' '))
      lines += batch.iso_originating_currency_code
      lines += batch.iso_destination_currency_code
      lines += batch.effective_entry_date.strftime('%y%m%d')
      lines += "#{sd}1"
      lines += originating_dfi.rjust(8, '0')
      lines += "#{batch_number.to_s.rjust(7, '0')}"
      lines += "\r\n"

      batch.entries.where(ach_re_batch_id: ach_re_batch.id).order(trace_number: :asc).each do |entry|
        result = batch.iat_entry_process(entry.id)
        lines += result[0]
        batch_entry_addenda_count += result[1]
        if entry.transaction_type == "CREDIT"
          credit_check += entry.amount_cents
        elsif entry.transaction_type == "DEBIT"
          debit_check += entry.amount_cents
        end
      end
    else
      lines = originator.name.nil? ? "5#{scc}#{' ' * 16}" : "5#{scc}#{originator.name.encode(Encoding.find('ASCII'), { :invalid => :replace, :undef => :replace, :replace => ' ' }).ljust(16, ' ')}"
      lines += (batch.company_discretionary_data.nil? ? ' ' * 20 : batch.company_discretionary_data.encode(Encoding.find('ASCII'), { :invalid => :replace, :undef => :replace, :replace => ' ' }).ljust(20, ' '))
      lines += (batch.company_identification.nil? ? ' ' * 10 : batch.company_identification.encode(Encoding.find('ASCII'), { :invalid => :replace, :undef => :replace, :replace => ' ' }).ljust(10, ' '))
      lines += sec
      lines += (batch.company_entry_description.nil? ? ' ' * 10 : batch.company_entry_description.encode(Encoding.find('ASCII'), { :invalid => :replace, :undef => :replace, :replace => ' ' }).ljust(10, ' '))
      lines += (batch.company_descriptive_date.nil? ? ' ' * 6 : batch.company_descriptive_date.ljust(6, ' '))
      lines += batch.effective_entry_date.strftime('%y%m%d')
      lines += "#{sd}1"
      lines += originating_dfi.rjust(8, '0')
      lines += "#{batch_number.to_s.rjust(7, '0')}"
      lines += "\r\n"

      batch.entries.where(ach_re_batch: ach_re_batch).order(trace_number: :asc).each do |entry|
        result = batch.entry_process(entry.id)
        lines += result[0]
        batch_entry_addenda_count += result[1]
        if entry.transaction_type == "CREDIT"
          credit_check += entry.amount_cents
        elsif entry.transaction_type == "DEBIT"
          debit_check += entry.amount_cents
        end
      end
    end

    entry_hash = AchAccount.joins(:entries).where(entries: { ach_re_batch_id: ach_re_batch.id, batch_id: batch.id }).sum(:routing_check)

    lines += "8#{scc}"
    lines += batch_entry_addenda_count.to_s.rjust(6, '0')
    lines += entry_hash.to_s.last(10).rjust(10, '0')
    lines += Entry.where(transaction_type: "DEBIT",
                         batch_id: batch.id,
                         ach_re_batch: ach_re_batch)
                  .sum(:amount_cents).to_s.rjust(12, '0')
    lines += Entry.where(transaction_type: "CREDIT",
                         batch_id: batch.id,
                         ach_re_batch: ach_re_batch)
                  .sum(:amount_cents).to_s.rjust(12, '0')
    lines += (batch.company_identification.nil? ? ' ' * 10 : batch.company_identification.ljust(10, ' '))
    lines += ' ' * 25
    lines += batch.originating_dfi_identification.rjust(8, ' ')
    lines += batch_number.to_s.rjust(7, '0')
    lines += "\r\n"

    batch = { lines: lines, batch_number: batch_number, credit_check: credit_check, debit_check: debit_check, entry_addenda_count: batch_entry_addenda_count }
    $redis.set("Rebatch ID: #{ach_re_batch.id}, Batch ID: #{batch_id}", batch.to_json)
    ach_re_batch_batch.update processed: true, batch_number: batch_number
  end

  def rebatch_assemble(ach_re_batch_id, file_count)
    ach_re_batch = AchReBatch.find_by_id(ach_re_batch_id)
    if ach_re_batch.ach_re_batch_batches.where(processed: false).any?
      ach_re_batch.entries.update_all(transmitted: false)
      ach_re_batch.ach_re_batch_batches.update_all(processed: false)
      AchReBatchAssembler.assemble(ach_re_batch_id, file_count)
    else
      line = "101#{ach_re_batch.ach_file.immediate_destination.rjust(10, ' ')}#{ach_re_batch.ach_file.immediate_origin.rjust(10, ' ')}#{Time.now.strftime('%y%m%d%H%M')}A094101#{ach_re_batch.ach_file.immediate_destination_name.ljust(23, ' ')}#{ach_re_batch.ach_file.immediate_origin_name.ljust(23, ' ')}#{ach_re_batch.ach_file.reference_code.ljust(8, ' ')}"
      lines = line.ljust(94, ' ')
      lines += "\r\n"

      credit_check = 0
      debit_check = 0
      entry_addenda_count = 0

      error_count = 0
      errors = []
      ach_re_batch.ach_re_batch_batches.order(batch_number: :asc).each do |rebatch|
        raw = $redis.get("Rebatch ID: #{ach_re_batch.id}, Batch ID: #{rebatch.batch.id}")
        batch = JSON.parse(raw) if raw.present?
        if error_count == 0 && batch.present?
          lines << batch["lines"]
          credit_check += batch["credit_check"].to_i
          debit_check += batch["debit_check"].to_i
          entry_addenda_count += batch["entry_addenda_count"].to_i
          $redis.del("Rebatch ID: #{ach_re_batch.id}, Batch ID: #{rebatch.batch.id}")
        else
          errors << "Rebatch ID: #{ach_re_batch.id}, Batch ID: #{rebatch.batch.id} missing"
          error_count += 1
        end
      end

      file_entry_hash = AchAccount.joins(:entries).where(entries: { ach_re_batch_id: ach_re_batch.id }).sum(:routing_check)

      line = "9"
      line += ach_re_batch.batch_count.to_s.rjust(6, '0')
      line += ((lines.lines.count.to_f + 1) / 10).ceil.to_s.rjust(6, '0')
      line += entry_addenda_count.to_s.rjust(8, '0')
      line += file_entry_hash.to_s.last(10).rjust(10, '0')
      line += ach_re_batch.debits_cents.to_s.rjust(12, '0')
      line += ach_re_batch.credits_cents.to_s.rjust(12, '0')
      line += ' ' * 39
      line = line.ljust(94, ' ')
      line += "\r\n"
      lines << line

      needed_nines = 10 - lines.lines.count.to_s.split(//).last(1)[0].to_i

      x = 0

      while x < needed_nines
        lines += "9" * 94
        lines += "\r\n"
        x += 1
      end

      length_problems = 0
      bad_lines = []
      lines.each_line do |line|
        unless line.length == 96
          length_problems += 1
          bad_lines << line
         end
      end

      if error_count == 0 && length_problems == 0 && credit_check == ach_re_batch.credits_cents && debit_check == ach_re_batch.debits_cents
        name = ach_re_batch.ach_file.name + "_" + file_count
        path = "tmp/" + name
        File.open(path, "w+") do |f|
          f.write(lines)
          ach_re_batch.update data_file: DataFile.create(file: f, data_file_type: DataFileType.find_or_create_by(name: "ACH ReBatch"))
        end
        ach_re_batch.import(File.open(path), name)
        File.delete(path)

        company_users = User.where(company: ach_re_batch.ach_file.company, wire_email: true)
        bank_users = User.where(bank: ach_re_batch.ach_file.company.bank, wire_email: true)

        company_users.each do |user|
          user.send_wire_email(ach_re_batch.id)
        end

        bank_users.each do |user|
          user.send_wire_email(ach_re_batch.id)
        end

        ach_re_batch.ach_file.company.send_wire_email(ach_re_batch.id) if ach_re_batch.ach_file.company.wires_email
        ach_re_batch.ach_file.bank.send_wire_email(ach_re_batch.id) if ach_re_batch.ach_file.bank.wires_email
      else
        user = User.find_by(email: "XXX@XXXXX.com")
        message = "Line lengths or Funds Totals not correct on ReBatch ID: #{ach_re_batch.id}\r\n"
        message += "Entries individual credit cents: #{credit_check}\r\n"
        message += "Entries individual debit cents: #{debit_check}\r\n"
        message += "Entries ReBatch sum credit cents: #{ach_re_batch.credits_cents}\r\n"
        message += "Entries ReBatch sum debit cents: #{ach_re_batch.debits_cents}\r\n"
        message += "Line length problems: #{length_problems}\r\n"
        message += "Bad Lines:\r\n"
        bad_lines.each do |line|
          message += line
        end
        message += "\r\n"
        message += "Batch missing from Redis: #{errors}"
        user.send_notification_email(message)
      end
    end
  end
end


```

**Comment by mperham** [comment]
The initializer is where you activate various optional Sidekiq Pro/Enterprise features.  If you don't have an initializer, that means you aren't enabling the reliability features.  I would immediately turn those on unless you have a compelling reason not to:

https://github.com/mperham/sidekiq/wiki/Reliability#setup

It's possible you are losing jobs somehow that the reliability features will save.

**Comment by DaveWhitmer** [comment]
Well, I guess I should have read the PRO directions more closely!  Trying that right now!

**Comment by mperham** [comment]
The only other really sketchy part of the code that I see there is the use of `$redis`.  Use of a single global connection is not a good idea in a heavily threaded system like Sidekiq.  It's easy to create and use your own connection pool with this migration path:

https://github.com/mperham/connection_pool/#migrating-to-a-connection-pool

**Comment by DaveWhitmer** [comment]
Thanks so much.  I am trying both.  The connection pool does seem to slow down Sidekiq a lot though.    I see that is mentioned in relation to storing values in redis, but not how to overcome this when using Sidekiq.  Is there something obvious I am missing?

**Comment by mperham** [comment]
By default, Sidekiq fetches N jobs in parallel.  super_fetch fetches one job at a time, so network round trip matters much more.

**Comment by DaveWhitmer** [comment]
I used this for my initializers:
sidekiq.rb:
```
Sidekiq::Client.reliable_push! unless Rails.env.test?

Sidekiq.configure_server do |config|
  config.reliable!
end
```
redis.rb:
`$redis = Redis::Namespace.new("default_redis", :redis => Redis.new)`
when I change redis.rb to this:
`$redis = ConnectionPool::Wrapper.new(size: 100, timeout: 3) { Redis.connect }
`

I have tried several pool sizes both in development and moved it to staging.  When I try using the Connection Pool it has about a 2 order of magnitude impact to performance.  Am I doing something wrong?

Good news is that the production system is now using the sidekiq initializer and at least this morning's file had no issues.  Take a week or two, before I'll breath easier about it though.  Thanks for all the help on that, I'm hopeful you have solved that problem for me.

**Comment by DaveWhitmer** [comment]
**Update:**  Once I noticed the reliable_fetch is deprecated message when I started Sidekiq, I switched the initializers to:
redis.rb:
`$redis = ConnectionPool::Wrapper.new(size: 100, timeout: 3) { Redis.connect }`
sidekiq.rb:
```
Sidekiq::Client.reliable_push! unless Rails.env.test?

Sidekiq.configure_server do |config|
	config.super_fetch!
	config.reliable_scheduler!
end
```

It runs MUCH better than I saw before, but I see what you mean by not being able to keep the thread count perfectly full on my development machine.  I'll see how it acts when I move it to staging, but I can certainly take a minor hit on performance for a huge increase in reliability.  I'll update on staging performance and production reliability when I can quantify it.

**Comment by mperham** [comment]
Change Redis.connect to Redis.new.

**Comment by DaveWhitmer** [comment]
Unfortunately, similar performance to Redis.connect.

**Comment by mperham** [comment]
Redis.connect maps to a single, "default" global TCP socket.  Redis.new maps to a new TCP socket for each instance.

Thanks for pointing out the quirk in `reliable!`, I updated the wiki to explicitly activate super_fetch.

**Comment by DaveWhitmer** [comment]
I believe this issue is now resolved.  I will open a new issue if it comes back.  Thank you!

**Comment by DaveWhitmer** [comment]
Unfortunately, I had two jobs this morning that did not complete.  I have a flag in the database that marks them as completed, so I can see that they were not completed, but they were started.  

The parent batch screenshot is attached, you can see nothing is in the queue and nothing is busy.  I was able to manually rerun the parent job and everything processed on the second try.  So I know there is not an error in the code.  Everything has been working fine for 1-2 weeks since I last addressed this issue.  

My initializer:
```
Sidekiq::Client.reliable_push! unless Rails.env.test?

Sidekiq.configure_server do |config|
	config.super_fetch!
	config.reliable_scheduler!
end
```

My redis intializer:
`$redis = ConnectionPool::Wrapper.new(size: 100, timeout: 3) { Redis.new }`

![screenshot 2018-06-19 07 25 18](https://user-images.githubusercontent.com/4665368/41594884-f0d60f16-7392-11e8-951a-57fd5b708605.png)

**Comment by DaveWhitmer** [comment]
Hit wrong button.  Now hitting Reopen.  Please see above comment.

**Comment by mperham** [comment]
I need a lot more context.  Go through your logs and determine why that job stopped processing and where it went.

**Comment by DaveWhitmer** [comment]
I am including the relevant portion of the logs for you.   I cannot seem to find a single job that was started and does not have a done entry.  There is something strange that occurs at time 04:22:12.  There is a log entry:
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: SuperFetch found 1 working queues with no orphaned jobs

So not sure what to make of it.  You can see from the screenshot in the above comment that the batch can see the jobs are there.  There's no errors in the log.  But there is this event that at 04:22:12, but then the jobs don't get picked back up and rerun after that or completed back to the parent batch as being completed or successful.  

Any help to point me in the right direction is so much appreciated!   It appears like maybe this is related to pg-bouncer?  It seems like this happens so intermittently, every couple weeks with hundreds of thousands of jobs processed in between the occurrences with no problem.

Please let me know what other information may be of assistance.

[log.txt](https://github.com/mperham/sidekiq/files/2116052/log.txt)

**Comment by DaveWhitmer** [comment]
Thanks for the hint that the JID's are on the batch page.  So much easier to track down.  Only references in my logs are to the jobs starting which makes sense from what the database shows.  No other references in the logs to either JID.  The weirdness in the logs is included below but is the same as before.

The only log entries around these are just the 812 out of 814 companion jobs that completed.

![screenshot 2018-06-19 15 30 14](https://user-images.githubusercontent.com/4665368/41620225-0b26ecfc-73d7-11e8-968a-dd275330c4ff.png)


`Jun 19 04:21:35 fcb-production app/smallworker.2:  37 TID-1ermi8 AchReBatchAssembler JID-9010f1b65e4f0cd686ff30a7 BID-SpHQIs6fYrLwmg INFO: start `
`Jun 19 04:21:48 fcb-production app/smallworker.2:  37 TID-1erlfc AchReBatchAssembler JID-5849d29f01717f9816466d69 BID-SpHQIs6fYrLwmg INFO: start`

```
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.221 38 LOG C-0x16577b8: db1/u28ritr1jnd5se@127.0.0.1:56888 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1657630: db1/u28ritr1jnd5se@127.0.0.1:56886 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x16574a8: db1/u28ritr1jnd5se@127.0.0.1:56884 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1657320: db1/u28ritr1jnd5se@127.0.0.1:56882 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1657198: db1/u28ritr1jnd5se@127.0.0.1:56880 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1657010: db1/u28ritr1jnd5se@127.0.0.1:56878 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1656d00: db1/u28ritr1jnd5se@127.0.0.1:56876 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: bin/start-pgbouncer-stunnel: line 136:    37 Killed                  "$@" 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.222 38 LOG C-0x1656b78: db1/u28ritr1jnd5se@127.0.0.1:56874 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x16569f0: db1/u28ritr1jnd5se@127.0.0.1:56872 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1656868: db1/u28ritr1jnd5se@127.0.0.1:56870 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1656e88: db1/u28ritr1jnd5se@127.0.0.1:56868 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x16566e0: db1/u28ritr1jnd5se@127.0.0.1:56866 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1656558: db1/u28ritr1jnd5se@127.0.0.1:56864 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x16563d0: db1/u28ritr1jnd5se@127.0.0.1:56862 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1656248: db1/u28ritr1jnd5se@127.0.0.1:56860 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x16560c0: db1/u28ritr1jnd5se@127.0.0.1:56858 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655f38: db1/u28ritr1jnd5se@127.0.0.1:56856 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655db0: db1/u28ritr1jnd5se@127.0.0.1:56854 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655c28: db1/u28ritr1jnd5se@127.0.0.1:56852 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655aa0: db1/u28ritr1jnd5se@127.0.0.1:56850 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655918: db1/u28ritr1jnd5se@127.0.0.1:56848 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.223 38 LOG C-0x1655790: db1/u28ritr1jnd5se@127.0.0.1:56846 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.224 38 LOG C-0x1655608: db1/u28ritr1jnd5se@127.0.0.1:56842 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.224 38 LOG C-0x1655480: db1/u28ritr1jnd5se@127.0.0.1:56840 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.224 38 LOG C-0x16552f8: db1/u28ritr1jnd5se@127.0.0.1:56836 closing because: client unexpected eof (age=86881) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.224 38 LOG C-0x1655170: db1/u28ritr1jnd5se@127.0.0.1:56824 closing because: client unexpected eof (age=86882) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.226 38 LOG Got SIGTERM, ignoring! 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=exit process=app 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.227 38 LOG Got SIGINT, shutting down 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.227 38 LOG S-0x1659e20: db1/u28ritr1jnd5se@unix:6101 closing because: pause mode (age=2343) 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018.06.19 11:22:12 LOG5[23]: Connection closed: 35949908 byte(s) sent to TLS, 198921513 byte(s) sent to socket 
Jun 19 04:22:12 fcb-production app/smallworker.2: 2018-06-19 11:22:12.258 38 LOG Got SIGTERM, ignoring! 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=app-end 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=kill-app pid=33 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=wait-app pid=33 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=kill-aux name=pgbouncer pid=32 signal=SIGINT 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=kill-aux name=stunnel pid=31 signal=SIGINT 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=stunnel-end 
Jun 19 04:22:12 fcb-production app/smallworker.2: buildpack=pgbouncer at=pgbouncer-end 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=pgbouncer-enabled 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=config-gen-start 
Jun 19 04:22:21 fcb-production app/smallworker.2: Setting DATABASE_URL_PGBOUNCER config var 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=config-gen-end 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=config-gen-override DATABASE_URL 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=stunnel-launched pid=30 signal=SIGINT 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=stunnel-start 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=pgbouncer-launched pid=31 signal=SIGINT 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=pgbouncer-start 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=app-launched pid=33 
Jun 19 04:22:21 fcb-production app/smallworker.2: buildpack=pgbouncer at=app-start 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018-06-19 11:22:21.229 34 LOG File descriptor limit: 10000 (H:10000), max_client_conn: 100, max fds possible: 111 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018-06-19 11:22:21.232 34 LOG listening on 127.0.0.1:6000 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018-06-19 11:22:21.232 34 LOG listening on unix:/tmp/.s.PGSQL.6000 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018-06-19 11:22:21.233 34 LOG process up: pgbouncer 1.7.2, libevent 2.0.21-stable (epoll), adns: evdns2, tls: OpenSSL 1.0.2g  1 Mar 2016 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: stunnel 5.40 on x86_64-unknown-linux-gnu platform 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: Compiled/running with OpenSSL 1.0.2g  1 Mar 2016 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: Threading:PTHREAD Sockets:POLL,IPv6,SYSTEMD TLS:ENGINE,FIPS,OCSP,PSK,SNI Auth:LIBWRAP 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: Reading configuration from file /app/vendor/stunnel/stunnel-pgbouncer.conf 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: UTF-8 byte order mark not detected 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: FIPS mode disabled 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG4[ui]: Service [DATABASE_URL] needs authentication to prevent MITM attacks 
Jun 19 04:22:21 fcb-production app/smallworker.2: 2018.06.19 11:22:21 LOG5[ui]: Configuration successful 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: [Sidetiq] Sidetiq v0.7.2 - Copyright (c) 2012-2013, Tobias Svensson <tob@tobiassvensson.co.uk> 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: [Sidetiq] Sidetiq is covered by the 3-clause BSD license. 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: [Sidetiq] See LICENSE and http://opensource.org/licenses/BSD-3-Clause for licensing details. 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: [Sidetiq] Sidetiq::Supervisor start 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-w71nw INFO: [Sidetiq] Sidetiq::Actor::Clock id: 54078340 initialize 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-8c5kk INFO: [Sidetiq] Sidetiq::Actor::Handler id: 70046722320180 initialize 
Jun 19 04:22:24 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Sidetiq::Actor::Handler id: 70046722321820 initialize 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Starting the New Relic agent version 5.1.0.344 in "production" environment. 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : To prevent agent startup add a NEW_RELIC_AGENT_ENABLED=false environment variable or modify the "production" section of your newrelic.yml. 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Reading configuration from config/newrelic.yml (/app) 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Environment: production 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Dispatcher: sidekiq 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Application: FCB 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Rails 3+ middleware instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing deferred Rack::Builder instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Rack::Builder middleware instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing http.rb instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Net instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Redis Instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing middleware-based Excon instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing ActiveJob instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Sidekiq instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing ActiveRecord 5 instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Rails 5 Controller instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Rails 5 Action Cable instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Installing Rails 5 view instrumentation 
Jun 19 04:22:25 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:25 +0000 smallworker.2 (39)] INFO : Finished instrumentation 
Jun 19 04:22:26 fcb-production app/smallworker.2: 2018-06-19 11:22:26.541 34 LOG C-0x2365170: db1/u28ritr1jnd5se@127.0.0.1:60618 login attempt: db=db1 user=u28ritr1jnd5se tls=no 
Jun 19 04:22:26 fcb-production app/smallworker.2: 2018-06-19 11:22:26.541 34 LOG S-0x2369e20: db1/u28ritr1jnd5se@unix:6101 new connection to server 
Jun 19 04:22:26 fcb-production app/smallworker.2: 2018.06.19 11:22:26 LOG5[0]: Service [DATABASE_URL] accepted connection from unnamed socket 
Jun 19 04:22:26 fcb-production app/smallworker.2: 2018.06.19 11:22:26 LOG5[0]: s_connect: connected 10.1.3.135:5432 
Jun 19 04:22:26 fcb-production app/smallworker.2: 2018.06.19 11:22:26 LOG5[0]: Service [DATABASE_URL] connected remote server from 10.0.138.40:48254 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: Booting Sidekiq 5.0.0 with redis options {:url=>"redis://fcb-production-redis.ggvxla.ng.0001.use1.cache.amazonaws.com:6379"} 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: Running in ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux] 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: Sidekiq Pro 3.7.1, commercially licensed.  Thanks for your support! 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: Starting processing, hit Ctrl-C to stop 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-1k1pcg INFO: SuperFetch activated 
Jun 19 04:22:27 fcb-production app/smallworker.2: 39 TID-8e9qs INFO: SuperFetch found 1 working queues with no orphaned jobs 
Jun 19 04:22:28 fcb-production app/smallworker.2: ** [NewRelic][2018-06-19 11:22:27 +0000 smallworker.2 (39)] INFO : Reporting to: https://rpm.newrelic.com/accounts/1976947/applications/135228719 
Jun 19 04:23:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: FastWorker (at: 1529407440.0) (last: 1529407380.0) 
Jun 19 04:23:21 fcb-production app/smallworker.2: 2018-06-19 11:23:21.232 34 LOG Stats: 0 req/s, in 149 b/s, out 818 b/s,query 1572 us 
Jun 19 04:24:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: FastWorker (at: 1529407500.0) (last: 1529407440.0) 
Jun 19 04:24:21 fcb-production app/smallworker.2: 2018-06-19 11:24:21.233 34 LOG Stats: 0 req/s, in 0 b/s, out 0 b/s,query 0 us 
Jun 19 04:25:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: FastWorker (at: 1529407560.0) (last: 1529407500.0) 
Jun 19 04:25:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: MediumWorker (at: 1529407800.0) (last: 1529407500.0) 
Jun 19 04:25:21 fcb-production app/smallworker.2: 2018-06-19 11:25:21.235 34 LOG Stats: 0 req/s, in 0 b/s, out 0 b/s,query 0 us 
Jun 19 04:26:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: FastWorker (at: 1529407620.0) (last: 1529407560.0) 
Jun 19 04:26:14 fcb-production app/smallworker.2: 2018-06-19 11:26:14.012 34 LOG C-0x23652f8: db1/u28ritr1jnd5se@127.0.0.1:60780 login attempt: db=db1 user=u28ritr1jnd5se tls=no 
Jun 19 04:26:14 fcb-production app/smallworker.2: 39 TID-1j9msg FastWorker JID-2e26502f0b9049fbf97b001f INFO: start 
Jun 19 04:26:14 fcb-production app/smallworker.2: 39 TID-1j9mko ScaleWorker JID-5615b234bc53140cc977e455 INFO: start 
Jun 19 04:26:14 fcb-production app/smallworker.2: 39 TID-1j9mko ScaleWorker JID-5615b234bc53140cc977e455 INFO: done: 0.115 sec 
Jun 19 04:26:14 fcb-production app/smallworker.2: 39 TID-1j9msg FastWorker JID-2e26502f0b9049fbf97b001f INFO: done: 0.334 sec 
Jun 19 04:26:21 fcb-production app/smallworker.2: 2018-06-19 11:26:21.235 34 LOG Stats: 0 req/s, in 41 b/s, out 440 b/s,query 21406 us 
Jun 19 04:27:00 fcb-production app/smallworker.2: 39 TID-w76p0 INFO: [Sidetiq] Enqueue: FastWorker (at: 1529407680.0) (last: 1529407620.0) 
Jun 19 04:27:21 fcb-production app/smallworker.2: 2018-06-19 11:27:21.236 34 LOG Stats: 0 req/s, in 0 b/s, out 0 b/s,query 0 us 
```

**Comment by DaveWhitmer** [comment]
You are not going to believe this, but that batch just completed.  It generates an email and sends it to me when the batch completes and I got the email at 3:50PM EDT.  Eight hours later.  Total and complete shock here.  Any ideas?

**Comment by DaveWhitmer** [comment]
Ok.  I'm impressed that it went in and found the orphaned jobs and ran them, but is there anything that you can see that I could do to avoid this situation?  I hate to write a time-based job that runs later to rerun it if it is not completed.  That opens up a can of worms and puts me back in the situation where we wanted the functionality that Pro gave us with batches.  The jobs we are doing can't take hours to complete or unknowingly run twice.

**Comment by mperham** [comment]
Fix the crash that is orphaning the jobs.  The Reliability feature is a **backup**; you still have a bug to fix.

First of all, do this: `bundle up sidekiq-pro`.  You are still running old Sidekiq versions.  Next, upgrade from Ruby 2.3.3 to 2.3.7.  Finally, upgrade all of your native gems; the wiki shows you how to find them.

**Comment by DaveWhitmer** [comment]
Ok.  Thank you.  I will try those recommendations.  I very much appreciate the help.

**Comment by DaveWhitmer** [comment]
I wanted to come back and document the result even though the issue is closed.  It appears that the app running on Heroku was running close to its memory limit and near the end of this batch job when it would execute the callbacks it would spike above the memory limit. Since I run on a private space there is no temporary allowance the way there is in the common runtime.  So it would cause issues with pg-bouncer that was running on the dyno.  My resolution is two-fold:  1) Increase dyno size and 2) move pg-bouncer to the SQL server (Heroku beta feature) . This initially appears to have gotten rid of this problem.  

I don't think I would have gotten to this result so quickly without the help of mperham.  Thank you so much for streamlining the process for me and getting to the heart of the problem quickly.
